compiler/rustc_builtin_macros/src/env.rs

```diff
@@ -21,7 +21,12 @@ pub fn expand_option_env<'cx>(
     };
 
     let sp = cx.with_def_site_ctxt(sp);
-    let value = env::var(var.as_str()).ok().as_deref().map(Symbol::intern);
+
+    let injected_value =
+        cx.sess.opts.injected_env_vars.get(var.as_str()).map(|s| Symbol::intern(s));
+    let value =
+        injected_value.or_else(|| env::var(var.as_str()).ok().as_deref().map(Symbol::intern));
+
     cx.sess.parse_sess.env_depinfo.borrow_mut().insert((var, value));
```

Comment:
I think this should be skipped in case `-Zinjected-env-vars` mentions this env var. Otherwise changes to the actual env vars will cause a recompilation, even if they don't affect compilation at all.